packages/contracts/contracts/libs/v0.8.x/minter-libs/TokenHolderLib.sol

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+import "@openzeppelin-4.5/contracts/utils/structs/EnumerableSet.sol";
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Art Blocks Token Holder Library
+ * @notice This library provides a collection of functions for managing and
+ * interacting with holders of specific NFTs, in the context of purchase gating for
+ * Art Blocks projects. The primary features of this library include the ability
+ * to register and unregister NFT addresses, the allowance and removal of holders
+ * from specific project tokens, as well as the ability to check if a specific NFT is
+ * allowlisted.
+ * @author Art Blocks Inc.
+ */
+
+library TokenHolderLib {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    // Define one million to represent the base token ID for each project.
+    uint256 constant ONE_MILLION = 1_000_000;
+
+    /**
+     * @dev Allows holders of specific project tokens.
+     * @param allowedProjectHoldersMapping A mapping of the allowed holders for a specific project.
+     * @param _projectId Project ID to enable minting on.
+     * @param _coreContract The address of the core contract.
+     * @param _ownedNFTAddresses NFT core addresses of projects to be
+     * allowlisted. Indexes must align with `_ownedNFTProjectIds`.
+     * @param _ownedNFTProjectIds Project IDs on `_ownedNFTAddresses` whose
+     * holders shall be allowlisted to mint project `_projectId`. Indexes must
+     * align with `_ownedNFTAddresses`.
+     */
+    function allowHoldersOfProjects(
+        mapping(address => mapping(uint256 => mapping(address => mapping(uint256 => bool))))
+            storage allowedProjectHoldersMapping,
+        uint256 _projectId,
+        address _coreContract,
```

Comment:
These large mappings can become a bit difficult to understand.
WDYT about doing the following:
```
// in this library, add the following struct, and names space per-project in the minter contract
// (same as we do for the Merkle project config):
struct  HolderProjectConfig {
    // projects whose holders are allowed to purchase a token on `projectId`
    // ownedNFTAddress => ownedNFTProjectIds => bool
    mapping(address => mapping(uint256 => bool)) allowedProjectHolders
}

// then in this function, we pass the struct directly, and can 
// remove the `_projectId` and `_coreContract` args 
function allowHoldersOfProjects(
    HolderProjectConfig storage holderProjectConfig,
    address[] memory _ownedNFTAddresses,
    uint256[] memory _ownedNFTProjectIds
) internal {
    ...
}
```

The above pattern also enables the `_projectId` and `_coreContract` input args to be removed from `removeHoldersOfProjects`, `allowAndRemoveHoldersOfProjects`, and `isAllowlistedNFT`.

Reply:
This is a great suggestion – thank you @ryley-o